pkg/controllers/elasticsearch/cluster_control.go

```diff
@@ -121,12 +136,197 @@ func (e *defaultElasticsearchClusterControl) Sync(c *v1alpha1.ElasticsearchClust
 	}
 
 	// TODO: handle status
-	if err = e.nodePoolControl.Sync(c); err != nil {
+	if c.Status, err = e.configMapControl.Sync(c); err != nil {
+		e.recorder.Eventf(c, apiv1.EventTypeWarning, errorSync, messageErrorSyncConfigMap, err.Error())
+		return c.Status, err
+	}
+
+	state := &controllers.State{
+		Clientset:          e.kubeClient,
+		NavigatorClientset: e.navigatorClient,
+		Recorder:           e.recorder,
+		StatefulSetLister:  e.statefulSetLister,
+		ConfigMapLister:    e.configMapLister,
+		PilotLister:        e.pilotLister,
+		PodLister:          e.podLister,
+	}
+
+	// for each node pool
+	// - check if ConfigMap is up to date
+	//   - if not, we need to create one
+	//     - create configmap
+	//     - update ElasticsearchCluster.status.nodePool[].desiredConfigMap with current configmap name
+	// - check if statefulset uses correct configmap
+	//   - if not, update statefulset with configmap name
+	//     - if desiredStatefulSetHash != current hash for statefulset,
+	//     - update ElasticsearchCluster.status.nodePool[].desiredStatefulSetHash with new statefulset hash from status
+	nextAction, err := e.nextAction(c)
+	if err != nil {
 		e.recorder.Eventf(c, apiv1.EventTypeWarning, errorSync, messageErrorSyncNodePools, err.Error())
 		return c.Status, err
 	}
 
-	e.recorder.Eventf(c, apiv1.EventTypeNormal, successSync, messageSuccessSync)
+	if nextAction == nil {
+		err = e.reconcileNodePools(c)
+		return c.Status, err
+	}
+
+	if nextAction != nil {
+		glog.Infof("Executing action %q", nextAction.Name())
+		err := nextAction.Execute(state)
+		glog.Infof("Finished executing action %q", nextAction.Name())
+		if err != nil {
+			e.recorder.Eventf(c, apiv1.EventTypeWarning, "Err"+nextAction.Name(), messageErrorSyncNodePools, err.Error())
+			return c.Status, err
+		}
+	}
+
+	return c.Status, err
+}
+
+func (e *defaultElasticsearchClusterControl) nextAction(c *v1alpha1.ElasticsearchCluster) (controllers.Action, error) {
+	for _, np := range c.Spec.NodePools {
+		action, err := e.nextActionForNodePool(c, &np)
+		if err != nil {
+			return nil, err
+		}
+		if action != nil {
+			return action, nil
+		}
+	}
+	return nil, nil
+}
+
+func (e *defaultElasticsearchClusterControl) nextActionForNodePool(c *v1alpha1.ElasticsearchCluster, np *v1alpha1.ElasticsearchClusterNodePool) (controllers.Action, error) {
+	statefulSetName := util.NodePoolResourceName(c, np)
+	statefulSet, err := e.statefulSetLister.StatefulSets(c.Namespace).Get(statefulSetName)
+	// create the node pool if it does not exist
+	if k8sErrors.IsNotFound(err) {
+		return &actions.CreateNodePool{c, np}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
```

Comment:
What happens if someone accidentally deletes a Navigator statefulset during an ongoing version upgrade from ES v2 to ES v4? 
 * Navigator will attempt to recreate the statefulset, but 
 * with a pod template containing the new version and 
 * start the newer version of the pilot and 
 * start the new version of ES 
 * without starting the intermediate ES v3 pilot and waiting for it to complete data migration. 

Perhaps instead, the `CreateNodePool` action needs to look at the last reported `ESCluster.Status.Version` (if it exists) and start that instead.

Does a similar problem exist for an ongoing scale down?
If the statefulset is deleted, a new one will be recreated with the desired (smaller) replica count whereas it should probably be recreated with the last known good replica count and wait for the shards to be drained from the chosen node?

This all supposes that the PVs from the deleted statefulset would be re-used by the new statefulset. Is that how it works?

Reply:
Or perhaps in the case where a statefulset goes missing, navigator should just stop, log an error event and wait for a human to repair it.

Reply:
And then, what would those steps be?
If a CassandraCluster resource is accidentally deleted, you still have all the data on, now orphan, PVs.
How would you recreate the cluster and safely re-use the data on those PVs?

Reply:
> What happens if someone accidentally deletes a Navigator statefulset during an ongoing version upgrade from ES v2 to ES v4?

This scenario is definitely something we are susceptible to. I think some examination of how the Deployment controller behaves in this case would be useful. If it *does* manage to handle it gracefully, my guess is because it stores a lot of information about the upgrade/rolling update back on the Deployment resource itself. We can definitely do something similar here.. borrowing a lot of logic from the Deployment controller itself.

I'm thinking (/hoping) that this doesn't need to be a release blocker however, as this actions change is quite architectural and it'd be good to get Cassandra working in a similar way too.

> Does a similar problem exist for an ongoing scale down?

Yes

> This all supposes that the PVs from the deleted statefulset would be re-used by the new statefulset. Is that how it works?

Yep, that is how it works

> And then, what would those steps be?
> If a CassandraCluster resource is accidentally deleted, you still have all the data on, now orphan, PVs.
> How would you recreate the cluster and safely re-use the data on those PVs?

Creating a statefulset with the same name (and same PVC template) should do it 😄